pyisy/node_servers.py

```diff
@@ -0,0 +1,352 @@
+        result = await self.isy.conn.request(
+            self.isy.conn.compile_url([URL_PROFILE_NS, "0", "connection"]),
+            ok404=False,
+        )
+        if result is None:
+            return
+
+        try:
+            connections_xml = minidom.parseString(result)
+        except XML_ERRORS:
+            _LOGGER.error("%s while parsing Node Server connections", XML_PARSE_ERROR)
+            raise ISYResponseParseError(XML_PARSE_ERROR)
+
+        connections = connections_xml.getElementsByTagName(TAG_CONNECTION)
+        for connection in connections:
+            self._connections.append(
+                NodeServerConnection(
+                    slot=attr_from_element(connection, ATTR_PROFILE),
+                    enabled=attr_from_element(connection, TAG_ENABLED),
+                    name=value_from_xml(connection, TAG_NAME),
+                    ssl=value_from_xml(connection, TAG_SSL),
+                    sni=value_from_xml(connection, TAG_SNI),
+                    port=value_from_xml(connection, TAG_PORT),
+                    timeout=value_from_xml(connection, TAG_TIMEOUT),
+                    isy_user_num=value_from_xml(connection, TAG_ISY_USER_NUM),
+                    ip_base_url=value_from_xml(connection, TAG_IP_BASE_URL),
+                    ns_user=value_from_xml(connection, TAG_NS_USER),
+                )
+            )
+        _LOGGER.info("ISY updated node server connection info")
+
+    async def get_node_server_profiles(self):
+        """Retrieve the node server definition files from the ISY."""
+        node_server_file_list = await self.isy.conn.request(
+            self.isy.conn.compile_url([URL_PROFILE_NS, "0", "files"]), ok404=False
+        )
+
+        if node_server_file_list is None:
+            return
+
+        _LOGGER.debug("Parsing node server file list")
+
+        try:
+            file_list_xml = minidom.parseString(node_server_file_list)
+        except XML_ERRORS:
+            _LOGGER.error("%s while parsing Node Server files", XML_PARSE_ERROR)
+            raise ISYResponseParseError(XML_PARSE_ERROR)
+
+        file_list: List[str] = []
+
+        profiles = file_list_xml.getElementsByTagName(ATTR_PROFILE)
+        for profile in profiles:
+            slot = attr_from_element(profile, ATTR_ID)
+            directories = profile.getElementsByTagName(TAG_FILES)
+            for directory in directories:
+                dir_name = attr_from_element(directory, ATTR_DIR)
+                files = directory.getElementsByTagName(TAG_FILE)
+                for file in files:
+                    file_name = attr_from_element(file, TAG_NAME)
+                    file_list.append(f"{slot}/download/{dir_name}/{file_name}")
+
+        file_tasks = [
+            self.isy.conn.request(self.isy.conn.compile_url([URL_PROFILE_NS, file]))
+            for file in file_list
+        ]
+        file_contents: List[str] = await asyncio.gather(*file_tasks)
```

Comment:
@bdraco - Not trying to drag you into reviewing this, but I'm stumped and with the amount of testing you do, I'm wondering if you've seen something like this before.

Do you have any guesses as to why the `asyncio.Semaphore` limiter would be causing the ISY to only return a partial result when grabbing several files? Even after waiting 1, 5, 10, 15 seconds to retry?

Background: I'm trying to add full support for NodeServers including downloading the custom profile files to get the correct names and status enums for everything. When I added a 2nd test node server, the Polisy will choke on the 6th file download, and only return `'<?xml version="1.0" encoding="UTF-8"?>'`. If I wait to retry, I still get the same thing several (up to 15) seconds later. However, if I up the limit on the semaphore (since this is newer hardware), everything loads fine.